profiling/src/profile/mod.rs

```diff
@@ -1155,6 +1320,1224 @@ mod api_test {
+                "coco",
+                upscaling_info,
+            )
+            .expect_err("Invalid offset");
+    }
+
+    #[test]
+    fn test_add_rule_with_offset_out_of_bound_poisson_function3() {
+        let sample_types = vec![
+            api::ValueType {
+                r#type: "samples",
+                unit: "count",
+            },
+            api::ValueType {
+                r#type: "wall-time",
+                unit: "nanoseconds",
+            },
+            api::ValueType {
+                r#type: "cpu-time",
+                unit: "nanoseconds",
+            },
+        ];
+
+        let mut profile: Profile = Profile::builder().sample_types(sample_types).build();
+
+        // adding same offsets
+        let by_value_offsets: Vec<usize> = vec![0, 4];
+        let upscaling_info = UpscalingInfo::Poisson {
+            x: 1100,
+            y: 100,
+            sampling_distance: 1,
+        };
+        profile
+            .add_upscaling_rules(
+                by_value_offsets.as_slice(),
+                "my label",
+                "coco",
+                upscaling_info,
+            )
+            .expect_err("Invalid offset");
+    }
+
+    #[test]
+    fn test_fails_serialization_if_two_rules_collide_at_runtime() {
+        let sample_types = vec![
+            api::ValueType {
+                r#type: "samples",
+                unit: "count",
+            },
+            api::ValueType {
+                r#type: "wall-time",
+                unit: "nanoseconds",
+            },
+            api::ValueType {
+                r#type: "cpu-time",
+                unit: "nanoseconds",
+            },
+        ];
+
+        let mut profile: Profile = Profile::builder().sample_types(sample_types).build();
+
+        let id_label = api::Label {
+            key: "my label",
+            str: Some("coco"),
+            num: 10,
+            num_unit: None,
+        };
+
+        let sample1 = api::Sample {
+            locations: vec![],
+            values: vec![1, 10000, 42],
+            labels: vec![id_label],
+        };
+
+        profile.add(sample1).expect("add to success");
+
+        let mut value_offsets: Vec<usize> = vec![0];
+        let upscaling_info = UpscalingInfo::Proportional {
+            total_sampled: 2,
+            total_real: 4,
+        };
+        profile
+            .add_upscaling_rules(value_offsets.as_slice(), "", "", upscaling_info)
+            .expect("Rule added");
+
+        // a bylabel rule on the third offset
+        value_offsets.clear();
+        value_offsets.push(0);
+        let upscaling_info2 = UpscalingInfo::Proportional {
+            total_sampled: 10,
+            total_real: 20,
+        };
+        profile
+            .add_upscaling_rules(
+                value_offsets.as_slice(),
+                id_label.key,
+                id_label.str.unwrap(),
+                upscaling_info2,
+            )
+            .expect("Rule added");
```

Comment:
`add_upscaling_rules` should fail at this point since rules always collide (if the second rule applies, then first one will also apply).

Reply:
You cannot tell until it runs. Because, if you do not have a pprof row with this specific label, it's still valid

Reply:
When you check for rule overlap, you should also check for overlap for `by value` rules (I added a comment there too), then you would catch this.
You can argue that a pprof without this rule is valid, but the same argument could be used for not checking for overlap when adding rules (ie. if the overlapping rules do not apply, then pprof is valid).
My argument is that this rule should be rejected because if it applies, then it will overlap an existing by value rule.